src/security/builtin_plugins/cryptographic/src/crypto_transform.c

```diff
@@ -1630,93 +1271,79 @@ static DDS_Security_boolean encode_rtps_message_encrypt (
-  set_crypto_header(header, transform_kind, transform_id, session->id, session->init_vector_suffix);
+
+  /* copy the rtps header to the encryption buffer */
+  (void)add_rtps_header(&buffer, rtps_header);
+
+  header = add_crypto_header(&buffer, SMID_SRTPS_PREFIX, transform_kind, transform_id, session->id, session->init_vector_suffix);
 
   if (is_encryption_required(transform_kind))
   {
-    struct encrypted_data *encrypted;
-    size_t submsg_len = secure_body_plain_size + sizeof (uint32_t);
-    if (submsg_len > UINT16_MAX)
-      goto enc_rtps_fail_data;
+    struct crypto_body *body = add_crypto_body(&buffer, secure_body_plain_size);
 
-    /* add SEC_BODY submessage */
-    body = add_submessage(&data, SMID_SEC_BODY_KIND, flags, submsg_len);
-    encrypted = (struct encrypted_data *)(body + 1);
-    contents = encrypted->data;
+    encrypted_data.base = body->content.data;
+    encrypted_data.length = secure_body_plain_size;
 
     /* encrypt message */
-    /* FIXME: improve performance by not copying plain_rtps_message to a new buffer (crypto_cipher_encrypt_data should allow encrypting parts of a message) */
-    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, header->session_id, secure_body_plain._buffer, secure_body_plain_size, NULL, 0, contents, &payload_len, &hmac, ex))
+    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, &header->prefix.iv, 2, plain_data, &encrypted_data, &hmac, ex))
       goto enc_rtps_fail_data;
 
-    encrypted->length = ddsrt_toBE4u(payload_len);
-    if (payload_len + sizeof(encrypted->length) > secure_body_plain_size)
+    body->content.length = ddsrt_toBE4u((uint32_t)encrypted_data.length);
+    if (encrypted_data.length > secure_body_plain_size)
     {
-      size_t inc = payload_len + sizeof(encrypted->length) - secure_body_plain_size;
-      body->length = (uint16_t)(body->length + inc);
-      data._length += (uint32_t)inc;
+      size_t inc = (size_t)(encrypted_data.length - secure_body_plain_size);
+      body->header.octetsToNextHeader = (uint16_t)(body->header.octetsToNextHeader + inc);
+      crypto_buffer_expand(&buffer, inc);
     }
   }
   else if (is_authentication_required(transform_kind))
   {
+    unsigned char *ptr = crypto_buffer_append(&buffer, secure_body_plain_size);
     /* the transformation_kind indicates only indicates authentication the determine HMAC */
-    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, header->session_id, NULL, 0, secure_body_plain._buffer, secure_body_plain_size, NULL, NULL, &hmac, ex))
+    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, &header->prefix.iv, 2, plain_data, NULL, &hmac, ex))
       goto enc_rtps_fail_data;
 
     /* copy submessage */
-    memcpy(contents, secure_body_plain._buffer, secure_body_plain_size);
-    payload_len = secure_body_plain_size;
-    data._length += payload_len;
+    memcpy(ptr, plain_data[0].base, plain_data[0].length);
```

Comment:
But then, if one decides to not hard-code `2`, this ought to get a static assertion ...